File: BlotoutFoundation/BlotoutFoundationAll/BOFUtilities/AESEncryptionHelper/BOCrypt.m
//
//  BOCrypt.m
//  BlotoutFoundation
//
//  Created by Blotout on 27/07/19.
//  Copyright © 2019 Blotout. All rights reserved.
//

#import "BOCrypt.h"

#import "NSData+Base64.h"
#import "NSString+Base64.h"
#import "NSData+CommonCrypto.h"

@implementation BOCrypt

+ (NSString *)encrypt:(NSString *)message key:(NSString *)key iv:(NSString *)iv {
    NSData *encryptedData = [[message dataUsingEncoding:NSUTF8StringEncoding] AES256EncryptedDataUsingKey: [[key dataUsingEncoding:NSUTF8StringEncoding] SHA256Hash] iv: [[iv dataUsingEncoding:NSUTF8StringEncoding] SHA256Hash] error: nil];
  NSString *base64EncodedString = [NSString base64StringFromData:encryptedData length:[encryptedData length]];
  return base64EncodedString;
}

+ (NSString *)encryptData:(NSData *)data key:(NSString *)key iv:(NSString *)iv {
    NSData *encryptedData = [data AES256EncryptedDataUsingKey: [[key dataUsingEncoding:NSUTF8StringEncoding] SHA256Hash] iv: [[iv dataUsingEncoding:NSUTF8StringEncoding] SHA256Hash] error: nil];
  NSString *base64EncodedString = [NSString base64StringFromData:encryptedData length:[encryptedData length]];
  return base64EncodedString;
}

+ (NSString *)encryptDataWithoutHash:(NSData *)data key:(NSString *)key iv:(NSString *)iv {
    NSData *encryptedData = [data AES256EncryptedDataUsingKey: [key dataUsingEncoding:NSUTF8StringEncoding] iv: [iv dataUsingEncoding:NSUTF8StringEncoding] error: nil];
  NSString *base64EncodedString = [NSString base64StringFromData:encryptedData length:[encryptedData length]];
  return base64EncodedString;
}


+ (NSString *)decrypt:(NSString *)base64EncodedString key:(NSString *)key iv:(NSString *)iv {
  NSData *encryptedData = [NSData base64DataFromString:base64EncodedString];
  NSData *decryptedData = [encryptedData decryptedAES256DataUsingKey:[[key dataUsingEncoding:NSUTF8StringEncoding] SHA256Hash] iv: [[iv dataUsingEncoding:NSUTF8StringEncoding] SHA256Hash] error:nil];
  return [[NSString alloc] initWithData:decryptedData encoding:NSUTF8StringEncoding];
}

@end

File: BlotoutAnalytics/Foundation/BOFUtilities/AESEncryptionHelper/BOCrypt.swift
//
//  BOCrypt.swift
//  BlotoutAnalyticsSDK
//
//  Created by Poonam Tiwari on 31/03/22.
//

import Foundation
class BOCrypt {
    class func encrypt(_ message: String?, key: String?, iv: String?) -> String? {
        var encryptedData: Data? = nil
        do {
            encryptedData = try message?.data(using: .utf8)?.aes256EncryptedData(usingKey: key?.data(using: .utf8)?.sha256Hash(), iv: iv?.data(using: .utf8)?.sha256Hash(), error: nil)
        } catch {
        }
        let base64EncodedString = String.base64String(from: encryptedData, length: (encryptedData?.count ?? 0))
        return base64EncodedString
    }
    class func encryptData(_ data: Data?, key: String?, iv: String?) -> String? {
        var encryptedData: Data? = nil
        do {
            encryptedData = try data?.aes256EncryptedData(usingKey: key?.data(using: .utf8)?.sha256Hash(), iv: iv?.data(using: .utf8)?.sha256Hash(), error: nil)
        } catch {
        }
        let base64EncodedString = String.base64String(from: encryptedData, length: (encryptedData?.count ?? 0))
        return base64EncodedString
    }
    
    class func encryptAndReturn(_ data: Data?, key: String?, iv: String?) -> Data? {
        var encryptedData: Data? = nil
        do {
            encryptedData = try data?.aes256EncryptedData(usingKey: key?.data(using: .utf8)?.sha256Hash(), iv: iv?.data(using: .utf8)?.sha256Hash(), error: nil)
        } catch {
        }
        return encryptedData
    }
    
    class func encryptDataWithoutHash(_ data: Data?, key: String?, iv: String?) -> String? {
        var encryptedData: Data? = nil
        do {
            encryptedData = try data?.aes256EncryptedData(usingKey: key?.data(using: .utf8), iv: iv?.data(using: .utf8), error: nil)
        } catch {
        }
        let base64EncodedString = String.base64String(from: encryptedData, length: (encryptedData?.count ?? 0))
        return base64EncodedString
    }
    
    class func decrypt(_ base64EncodedString: String?, key: String?, iv: String?) -> String? {
        let encryptedData = Data.base64Data(from: base64EncodedString)
        var decryptedData: Data? = nil
        do {
            decryptedData = try encryptedData?.decryptedAES256Data(usingKey: key?.data(using: .utf8)?.sha256Hash(), iv: iv?.data(using: .utf8)?.sha256Hash())
        } catch {
        }
        if let decryptedData = decryptedData {
            return String(data: decryptedData, encoding: .utf8)
        }
        return nil
    }
    
    class func decryptAndReturn(_ data: Data?, key: String?, iv: String?) -> Data? {
        var decryptedData: Data? = nil
        do {
            decryptedData = try data?.decryptedAES256Data(usingKey: key?.data(using: .utf8)?.sha256Hash(), iv: iv?.data(using: .utf8)?.sha256Hash())
        } catch {
        }
        return decryptedData
    }
}
